File: lib/src/native_ad/native_ad_options.dart
/// Native options builder

import 'package:flutter/material.dart';

import '../../stack_appodeal_flutter.dart';

class NativeAdOptions with AppodealPlatformArguments {
  final _NativeAdType nativeAdType;
  final AdTitleConfig adTitleConfig;
  final AdAttributionConfig adAttributionConfig;
  final AdChoiceConfig adChoiceConfig;
  final AdIconConfig adIconConfig;
  final AdDescriptionConfig adDescriptionConfig;
  final AdActionButtonConfig adActionButtonConfig;
  final AdMediaConfig adMediaConfig;

  NativeAdOptions._({
    required this.nativeAdType,
    required this.adTitleConfig,
    required this.adAttributionConfig,
    required this.adChoiceConfig,
    required this.adIconConfig,
    required this.adDescriptionConfig,
    required this.adActionButtonConfig,
    required this.adMediaConfig,
  });

  /// Generates custom Native Ad View options
  static NativeAdOptions customOptions({
    AdTitleConfig? adTitleConfig,
    AdAttributionConfig? adAttributionConfig,
    AdChoiceConfig? adChoiceConfig,
    AdIconConfig? adIconConfig,
    AdDescriptionConfig? adDescriptionConfig,
    AdActionButtonConfig? adActionButtonConfig,
    AdMediaConfig? adMediaConfig,
  }) {
    return NativeAdOptions._(
      nativeAdType: _NativeAdType.custom,
      adTitleConfig: adTitleConfig ?? AdTitleConfig(),
      adAttributionConfig: adAttributionConfig ?? AdAttributionConfig(),
      adChoiceConfig: adChoiceConfig ?? AdChoiceConfig(),
      adIconConfig: adIconConfig ?? AdIconConfig(),
      adDescriptionConfig: adDescriptionConfig ?? AdDescriptionConfig(),
      adActionButtonConfig: adActionButtonConfig ?? AdActionButtonConfig(),
      adMediaConfig: adMediaConfig ?? AdMediaConfig(),
    );
  }

  /// Generates template Native Ad View options
  static NativeAdOptions _templateOptions({
    required _NativeAdType nativeAdType,
    int? adIconSize,
    int? adTitleFontSize,
    int? adActionButtonTextSize,
    int? adDescriptionFontSize,
    Color? adAttributionTextColor,
    Color? adAttributionBackgroundColor,
    AdChoicePosition? adChoicePosition,
    bool? adIconVisible,
    bool? adMediaVisible,
  }) {
    return NativeAdOptions._(
      nativeAdType: nativeAdType,
      adTitleConfig: AdTitleConfig(fontSize: adTitleFontSize ?? 16),
      adAttributionConfig: AdAttributionConfig(
          textColor: adAttributionTextColor ?? Colors.black,
          backgroundColor: adAttributionBackgroundColor ?? Colors.transparent),
      adChoiceConfig:
          AdChoiceConfig(position: adChoicePosition ?? AdChoicePosition.endTop),
      adIconConfig:
          AdIconConfig(visible: adIconVisible ?? true, size: adIconSize ?? 50),
      adDescriptionConfig:
          AdDescriptionConfig(fontSize: adDescriptionFontSize ?? 14),
      adActionButtonConfig:
          AdActionButtonConfig(fontSize: adActionButtonTextSize ?? 14),
      adMediaConfig: AdMediaConfig(visible: adMediaVisible ?? true),
    );
  }

  /// Generates Content Stream template Native Ad View options
  static NativeAdOptions contentStreamOptions({
    int? adTitleFontSize,
    int? adActionButtonTextSize,
    int? adDescriptionFontSize,
    Color? adAttributionTextColor,
    Color? adAttributionBackgroundColor,
    AdChoicePosition? adChoicePosition,
  }) {
    return NativeAdOptions._templateOptions(
      nativeAdType: _NativeAdType.contentStream,
      adIconSize: 90,
      adTitleFontSize: adTitleFontSize,
      adActionButtonTextSize: adActionButtonTextSize,
      adDescriptionFontSize: adDescriptionFontSize,
      adAttributionTextColor: adAttributionTextColor,
      adAttributionBackgroundColor: adAttributionBackgroundColor,
      adChoicePosition: adChoicePosition,
      adIconVisible: false,
      adMediaVisible: true,
    );
  }

  /// Generates App Wall template Native Ad View options
  static NativeAdOptions appWallOptions({
    int? adTitleFontSize,
    int? adActionButtonTextSize,
    int? adDescriptionFontSize,
    Color? adAttributionTextColor,
    Color? adAttributionBackgroundColor,
    AdChoicePosition? adChoicePosition,
  }) {
    return NativeAdOptions._templateOptions(
      nativeAdType: _NativeAdType.appWall,
      adIconSize: 70,
      adTitleFontSize: adTitleFontSize,
      adActionButtonTextSize: adActionButtonTextSize,
      adDescriptionFontSize: adDescriptionFontSize,
      adAttributionTextColor: adAttributionTextColor,
      adAttributionBackgroundColor: adAttributionBackgroundColor,
      adChoicePosition: adChoicePosition,
      adIconVisible: false,
      adMediaVisible: false,
    );
  }

  /// Generates News Feed template Native Ad View options
  static NativeAdOptions newsFeedOptions({
    int? adTitleFontSize,
    int? adActionButtonTextSize,
    int? adDescriptionFontSize,
    Color? adAttributionTextColor,
    Color? adAttributionBackgroundColor,
    AdChoicePosition? adChoicePosition,
  }) {
    return NativeAdOptions._templateOptions(
      nativeAdType: _NativeAdType.newsFeed,
      adIconSize: 50,
      adTitleFontSize: adTitleFontSize,
      adActionButtonTextSize: adActionButtonTextSize,
      adDescriptionFontSize: adDescriptionFontSize,
      adAttributionTextColor: adAttributionTextColor,
      adAttributionBackgroundColor: adAttributionBackgroundColor,
      adChoicePosition: adChoicePosition,
      adIconVisible: false,
      adMediaVisible: false,
    );
  }

  double getWidgetHeight(BuildContext context) {
    double width = getWidgetWidth(context);
    switch (nativeAdType) {
      case _NativeAdType.newsFeed:
        return width * 0.146;
      case _NativeAdType.appWall:
        return width * 0.195;
      case _NativeAdType.contentStream:
        return width * 0.764;
      default:
        return getAdHeight(context, width);
    }
  }

  double getWidgetWidth(BuildContext context) {
    return MediaQuery.of(context).size.width;
  }

  // TODO: two methods for get height
  double getAdHeight(BuildContext context, double width) {
    double height = 0;
    if (!adMediaConfig.visible && !adIconConfig.visible) {
      return 0;
    }

    if (adMediaConfig.visible) {
      height = width * 0.563; // 9:16
    }
    if (adIconConfig.visible) {
      height = height + adIconConfig.size;
    } else {
      height = height + 50; // height + adIconConfig._defaultSize;
    }
    return height.toDouble();
  }

  /// Convert to map to pass to NativeAdOptions in NativeAd
  @override
  Map<String, dynamic> get toMap => <String, dynamic>{
        'nativeAdType': nativeAdType.index,
        'adMediaConfig': adMediaConfig.toMap,
        'adTitleConfig': adTitleConfig.toMap,
        'adAttributionConfig': adAttributionConfig.toMap,
        'adChoiceConfig': adChoiceConfig.toMap,
        'adIconConfig': adIconConfig.toMap,
        'adDescriptionConfig': adDescriptionConfig.toMap,
        'adActionButtonConfig': adActionButtonConfig.toMap
      };
}

enum _NativeAdType { custom, contentStream, appWall, newsFeed }

File: android/src/main/kotlin/com/appodeal/appodeal_flutter/native_ad/NativeAdOptions.kt
package com.appodeal.appodeal_flutter.native_ad

import com.appodeal.appodeal_flutter.apdLog
import com.appodeal.appodeal_flutter.native_ad.models.AdActionButtonConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdAttributionConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdChoiceConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdDescriptionConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdIconConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdLayoutConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdMediaConfig
import com.appodeal.appodeal_flutter.native_ad.models.AdTitleConfig

class NativeAdOptions(
    val nativeAdViewType: NativeAdViewType,
    val adActionButtonConfig: AdActionButtonConfig,
    val adAttributionConfig: AdAttributionConfig,
    val adChoiceConfig: AdChoiceConfig,
    val adDescriptionConfig: AdDescriptionConfig,
    val adIconConfig: AdIconConfig,
//    val adLayoutConfig: AdLayoutConfig,
    val adMediaConfig: AdMediaConfig,
    val adTitleConfig: AdTitleConfig,
) {
    override fun toString(): String {
        return "NativeAdOptions(nativeAdViewType=$nativeAdViewType, adActionButtonConfig=$adActionButtonConfig, adAttributionConfig=$adAttributionConfig, adChoiceConfig=$adChoiceConfig, adDescriptionConfig=$adDescriptionConfig, adIconConfig=$adIconConfig, adMediaConfig=$adMediaConfig, adTitleConfig=$adTitleConfig)"
    }

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun toNativeAdOptions(args: Map<String, Any>): NativeAdOptions? = runCatching {
            apdLog("toNativeAdOptions: $args")
            val idxNativeAdViewType = args["nativeAdType"] as? Int
                ?: throw IllegalArgumentException("nativeAdType is required")
            NativeAdOptions(
                nativeAdViewType = NativeAdViewType.values()[idxNativeAdViewType],
                adActionButtonConfig = AdActionButtonConfig.toAdActionButtonConfig(args["adActionButtonConfig"] as Map<String, Any>),
                adAttributionConfig = AdAttributionConfig.toAdAttributionConfig(args["adAttributionConfig"] as Map<String, Any>),
                adChoiceConfig = AdChoiceConfig.toAdChoiceConfig(args["adChoiceConfig"] as Map<String, Any>),
                adDescriptionConfig = AdDescriptionConfig.toAdDescriptionConfig(args["adDescriptionConfig"] as Map<String, Any>),
                adIconConfig = AdIconConfig.toAdIconConfig(args["adIconConfig"] as Map<String, Any>),
//                adLayoutConfig = AdLayoutConfig.toAdLayoutConfig(args["adLayoutConfig"] as Map<String, Any>),
                adMediaConfig = AdMediaConfig.toAdMediaConfig(args["adMediaConfig"] as Map<String, Any>),
                adTitleConfig = AdTitleConfig.toAdTitleConfig(args["adTitleConfig"] as Map<String, Any>),
            ).also {
                apdLog("NativeAdOptions: $it")
            }
        }.getOrNull()
    }
}

//D/AppodealFlutterPlugin(31648): toNativeAdOptions: {adChoiceConfig={position=3}, adMediaConfig={margin=0, visible=false, position=0}, adDescriptionConfig={backgroundColor=#00000000, margin=0, fontSize=12, textColor=#FF000000}, adActionButtonConfig={backgroundColor=#00000000, margin=0, fontSize=14, radius=8, textColor=#FF000000}, adTitleConfig={backgroundColor=#00000000, margin=0, fontSize=14, textColor=#FF000000}, adAttributionConfig={backgroundColor=#FFFF6E40, margin=0, fontSize=12, textColor=#FF000000}, adIconConfig={margin=0, visible=false, size=50, position=0}, nativeAdType=3}
//D/AppodealFlutterPlugin(31648): toAdActionButtonConfig: {backgroundColor=#00000000, margin=0, fontSize=14, radius=8, textColor=#FF000000}
//D/AppodealFlutterPlugin(31648): toAdAttributionConfig: {backgroundColor=#FFFF6E40, margin=0, fontSize=12, textColor=#FF000000}
//D/AppodealFlutterPlugin(31648): toAdChoiceConfig: {position=3}
//D/AppodealFlutterPlugin(31648): toAdDescriptionConfig: {backgroundColor=#00000000, margin=0, fontSize=12, textColor=#FF000000}
//D/AppodealFlutterPlugin(31648): toAdIconConfig: {margin=0, visible=false, size=50, position=0}

enum class NativeAdViewType { Custom, ContentStream, AppWall, NewsFeed }